provision/tests/provisioners.sh

```diff
@@ -0,0 +1,142 @@
+    bash /srv/provision/provision-pre.sh
+  fi
+}
+
+function post_hook() {
+  # provision-post.sh
+  #
+  # acts as a post-hook to the default provisioning. Anything that should
+  # run after the shell commands laid out in provision.sh or provision-custom.sh should be
+  # put into this file. This provides a good opportunity to install additional packages
+  # without having to replace the entire default provisioning script.
+  if [[ -f "/srv/provision/provision-post.sh" ]]; then
+    VVV_LOG="post"
+    bash /srv/provision/provision-post.sh
+  fi
+}
+
+function dashboard() {
+  local dashboard_repo=$(get_config_value "dashboard.repo" "https://github.com/Varying-Vagrant-Vagrants/dashboard.git")
+  local dashboard_branch=$(get_config_value "dashboard.branch" "master")
+
+  VVV_LOG="dashboard"
+  bash /srv/provision/provision-dashboard.sh "${dashboard_repo}" "${dashboard_branch}"
+}
+
+function utility_sources() {
+  local name=()
+  local repo=()
+  local branch=()
+
+  local key="utility-sources"
+
+  local utilities=$(get_config_type "${key}")
+  if [[ "${utilities}" == "struct" ]]; then
+    utilities=($(get_config_keys "${key}"))
+  else
+    utilities=$(get_config_value "${key}")
+    if [[ ! -z "${utilities}" ]]; then
+      vvv_error "Malformed ${key} config"
+    fi
+    utilities=()
+  fi
+
+  containsElement "core" "${utilities}"
+  if [[ $? -ne 0 ]]; then
+    name+=("core")
+    repo+=("https://github.com/Varying-Vagrant-Vagrants/vvv-utilities.git")
+    branch+=("master")
+  fi
+
+  local utility
+  for utility in "${utilities[@]}"; do
+    type=$(get_config_type "${key}.${utility}")
+    name+=(${utility})
+    if [[ "${type}" == "str" ]]; then
+      repo+=($(get_config_value "${key}.${utility}"))
+      branch+=(master)
+    else
+      repo+=($(get_config_value "${key}.${utility}.repo"))
+      branch+=($(get_config_value "${key}.${utility}.branch" "master"))
+    fi
+  done
+
+  local i
+  for i in ${!name[@]}; do
+    VVV_LOG="utility-source-${name[$i]}"
+    bash /srv/provision/provision-utility-source.sh "${name[$i]}" "${repo[$i]}" "${branch[$i]}"
+  done
+}
+
+function provision_utilities() {
+  local groups=($(get_config_keys utilities))
+  local group
+  local utility
+  for group in ${groups[@]}; do
+    local utilities=($(get_config_values utilities."${group}"))
+    for utility in ${utilities[@]}; do
+      provision_utility "${group}" "${utility}"
+    done
+  done
+}
+
+function provision_utility() {
+  local group=$1
+  local utility=$2
+  VVV_LOG="utility-${group}-${utility}"
+  bash /srv/provision/provision-utility.sh "${group}" "${utility}"
+}
+
+function sites() {
+  local sites=($(get_config_keys sites))
+  local site
+  for site in ${sites[@]}; do
+    local skip_provisioning=$(get_config_value "sites.${site}.skip_provisioning" "False")
+    if [[ $skip_provisioning == "True" ]]; then
+      continue
+    fi
+    local repo=$(get_config_type "sites.${site}")
+    if [[ "${repo}" == "str" ]]; then
```

Comment:
I've forgotten where `str` fits in here 🤔 maybe i'm having brainfuzz

Reply:
This checks for the type of the key, if you have:

```
sites:
  my-site: "https://github.com/something/else"
```

then sites.my-site is a str, not an object, therefore that's the repo. If it's not a str, then you likely have the repo key inside that object.

Reply:
Oooh I didn't see `get_config_type`